polkadot/xcm/pallet-xcm/src/precompiles.rs

```diff
@@ -0,0 +1,136 @@
+
+use crate::{Config, VersionedLocation, VersionedXcm, Weight, WeightInfo};
+use alloc::vec::Vec;
+use codec::{DecodeAll, Encode};
+use core::{marker::PhantomData, num::NonZero};
+use pallet_revive::{
+	precompiles::{
+		alloy::{self, sol_types::SolValue},
+		AddressMatcher, Error, Ext, Precompile, RuntimeCosts,
+	},
+	Origin,
+};
+use tracing::error;
+use xcm_executor::traits::WeightBounds;
+
+alloy::sol!("src/precompiles/IXcm.sol");
+use IXcm::*;
+
+pub struct XcmPrecompile<T>(PhantomData<T>);
+
+impl<Runtime> Precompile for XcmPrecompile<Runtime>
+where
+	Runtime: crate::Config + pallet_revive::Config,
+{
+	type T = Runtime;
+	const MATCHER: AddressMatcher = AddressMatcher::Fixed(NonZero::new(10).unwrap());
+	const HAS_CONTRACT_INFO: bool = false;
+	type Interface = IXcm::IXcmCalls;
+
+	fn call(
+		_address: &[u8; 20],
+		input: &Self::Interface,
+		env: &mut impl Ext<T = Self::T>,
+	) -> Result<Vec<u8>, Error> {
+		let origin = env.caller();
+		let frame_origin = match origin {
+			Origin::Root => frame_system::RawOrigin::Root.into(),
+			Origin::Signed(account_id) =>
+				frame_system::RawOrigin::Signed(account_id.clone()).into(),
+		};
+
+		match input {
+			IXcmCalls::xcmSend(IXcm::xcmSendCall { destination, message }) => {
+				let weight = <Runtime as Config>::WeightInfo::send();
+				let _ = env.gas_meter_mut().charge(RuntimeCosts::Precompile(weight))?;
+				let final_destination = VersionedLocation::decode_all(&mut &destination[..])
+					.map_err(|error| {
+						error!(target: "Xcm", "XCM send failed: Invalid destination format {error:?}");
+						Error::Revert("XCM send failed: Invalid destination format".into())
+					})?;
+
+				let final_message =
+					VersionedXcm::<()>::decode_all(&mut &message[..]).map_err(|error| {
+						error!(target: "Xcm", "XCM send failed: Invalid message format {error:?}");
+						Error::Revert("XCM send failed: Invalid message format".into())
+					})?;
+
+				crate::Pallet::<Runtime>::send(
+					frame_origin,
+					final_destination.into(),
+					final_message.into(),
+				)
+				.map(|message_id| message_id.encode())
+				.map_err(|error| {
+					error!(
+						target: "Xcm",
+						"XCM send failed: destination or message format may be incompatible {error:?}"
+
+					);
+					Error::Revert(
+						"XCM send failed: destination or message format may be incompatible".into(),
+					)
+				})
+			},
+			IXcmCalls::xcmExecute(IXcm::xcmExecuteCall { message, weight }) => {
+				let weight = Weight::from_parts(weight.refTime, weight.proofSize);
+				let _ = env.gas_meter_mut().charge(RuntimeCosts::Precompile(weight))?;
+
+				let final_message =
+					VersionedXcm::decode_all(&mut &message[..]).map_err(|error| {
+						error!(target: "xcm", "XCM execute failed: Invalid message format {error:?}");
+						Error::Revert("Invalid message format".into())
+					})?;
+
+				crate::Pallet::<Runtime>::execute(frame_origin, final_message.into(), weight)
```

Comment:
We are missing a weight refund here.

Reply:
Please check if the solution is good

Reply:
No need to handle it twice. Use this function: https://paritytech.github.io/polkadot-sdk/master/frame_support/dispatch/fn.extract_actual_weight.html

Then you don't need to clone the value either.